# Python Library Complex Table PDF Parsing Feasibility Study

Comparision of various Python libraries and one Java library for parsing complex tables in PDF.  Arguably in order from worst to best.

The various libraries will have to be installed to run this notebook.

## pdfplumber

In [1]:
pdf_file = "OnePageOneTable.pdf"

In [2]:
from collections import namedtuple
import re

import pdfplumber
import pandas as pd

with pdfplumber.open(pdf_file) as f:
    for page in f.pages:
        text = page.extract_text()
        if text.__contains__("Consolidated Statements of Income"):
            print(text)
            print(page.extract_tables())

Consolidated Statements of Income
Barrick Gold Corporation Three months ended Six months ended
(in millions of United States dollars, except per share data) (Unaudited) June 30, June 30,
2015 2014 2015 2014
Revenue (notes 5 and 6) $ 2,231 $ 2,458 $ 4,476 $ 5,105
Costs and expenses (income)
Cost of sales (notes 5 and 7) 1,689 1,631 3,397 3,350
General and administrative expenses 70 82 137 185
Exploration, evaluation and project expenses (note 8) 9 7 1 05 183 205
Impairment charges (note 10B) 35 512 40 524
Loss on currency translation 33 31 31 110
Closed mine rehabilitation (19) 27 (11) 49
Loss (gain) on non-hedge derivatives (note 18D) 8 ( 44) 11 (65)
Other expense (note 10A) 32 17 14 36
Income before finance items and income taxes $ 286 $ 97 $ 674 $ 711
Finance items
Finance income 2 3 4 6
Finance costs (note 11) (194) (200) (390) ( 401)
Income (loss) before income taxes $ 94 $ (100) $ 288 $ 316
Income tax expense (note 12) (103) ( 123) (208) (412)
Net income (loss) $ (9) $ (223) $ 80 

## PyMuPDF

In [4]:
import sys, pathlib, pymupdf

with pymupdf.open(pdf_file) as doc:  # open document
    text = chr(12).join([page.get_text() for page in doc])
# write as a binary file to support non-ASCII characters
pathlib.Path(pdf_file + ".txt").write_bytes(text.encode())
myDoc = pymupdf.open(pdf_file)
# print(myDoc[0].get_text())
lines = myDoc[0].get_text('blocks')
for line in lines:
    print(line)

(57.2400016784668, 76.05886840820312, 337.3027038574219, 100.50143432617188, 'Consolidated Statements of Income\n', 0, 0)
(55.560001373291016, 114.79463195800781, 548.5789184570312, 125.43806457519531, 'Barrick Gold Corporation\nThree months ended\nSix months ended\n', 1, 0)
(55.560001373291016, 127.51466369628906, 548.502197265625, 149.74334716796875, '(in millions of United States dollars, except per share data) (Unaudited)\nJune 30,\nJune 30,\n2015\n \n2014\n2015\n2014\n', 2, 0)
(55.6783447265625, 167.4749298095703, 545.9047241210938, 251.5035400390625, 'Revenue (notes 5 and 6)\n2,231\n$     \n2,458\n$      \n4,476\n$   \n5,105\n$        \nCosts and expenses (income)\nCost of sales (notes 5 and 7)\n1,689\n       \n1,631\n          \n3,397\n     \n3,350\n         \nGeneral and administrative expenses \n70\n           \n82\n             \n137\n         \n185\n             \nExploration, evaluation and project expenses (note 8)\n97\n            \n105\n            \n183\n         \n205\

In [5]:
import fitz  # PyMuPDF

# Function to save the first page of a PDF
def save_first_page(input_pdf_path, output_pdf_path):
    # Open the input PDF file
    pdf_document = fitz.open(input_pdf_path)
    
    # Select the first page (index 0)
    first_page = pdf_document.load_page(0)  # Index starts at 0
    
    # Create a new PDF document to save the first page
    new_pdf = fitz.open()
    
    # Insert the first page into the new PDF
    new_pdf.insert_pdf(pdf_document, from_page=0, to_page=0)
    
    # Save the new PDF
    new_pdf.save(output_pdf_path)
    new_pdf.close()
    pdf_document.close()

# Example usage
input_pdf_path = pdf_file  # Replace with your input PDF path
output_pdf_path = "OnePageOneTable.pdf" # Replace with your output PDF path
save_first_page(input_pdf_path, output_pdf_path)


## pdfminer

In [8]:
from pdfminer.high_level import extract_text

text = extract_text(pdf_file, page_numbers=[0])
print(text)

Consolidated Statements of Income

Barrick Gold Corporation

(in millions of United States dollars, except per share data) (Unaudited)

Revenue (notes 5 and 6)
Costs and expenses (income)
Cost of sales (notes 5 and 7)
General and administrative expenses 
Exploration, evaluation and project expenses (note 8)
Impairment charges (note 10B)
Loss on currency translation
Closed mine rehabilitation
Loss (gain) on non-hedge derivatives (note 18D)
Other expense (note 10A)
Income before finance items and income taxes
Finance items
Finance income
Finance costs (note 11)
Income (loss) before income taxes
Income tax expense (note 12)
Net income (loss)
Attributable to:
Equity holders of Barrick Gold Corporation
Non-controlling interests (note 21)

Three months ended
June 30,
2014

2015

Six months ended
June 30,
2014

2015

$     

2,231

$      

2,458

$   

4,476

$        

5,105

1,689
70
97
35
33
(19)
8
32
286

$       

1,631
82
105
512
31
27
(44)
17
97

$           

3,397
137
183
40
31
(11)

In [9]:
from pdfminer.high_level import extract_pages
for page_layout in extract_pages(pdf_file, page_numbers=[0]):
    for element in page_layout:
        print(element)

<LTTextBoxHorizontal(0) 57.240,691.499,337.303,713.579 'Consolidated Statements of Income\n'>
<LTTextBoxHorizontal(1) 55.560,666.562,128.507,674.842 'Barrick Gold Corporation\n'>
<LTTextBoxHorizontal(2) 55.560,653.842,265.624,662.122 '(in millions of United States dollars, except per share data) (Unaudited)\n'>
<LTTextBoxHorizontal(3) 55.678,376.339,231.840,623.536 'Revenue (notes 5 and 6)\nCosts and expenses (income)\nCost of sales (notes 5 and 7)\nGeneral and administrative expenses \nExploration, evaluation and project expenses (note 8)\nImpairment charges (note 10B)\nLoss on currency translation\nClosed mine rehabilitation\nLoss (gain) on non-hedge derivatives (note 18D)\nOther expense (note 10A)\nIncome before finance items and income taxes\nFinance items\nFinance income\nFinance costs (note 11)\nIncome (loss) before income taxes\nIncome tax expense (note 12)\nNet income (loss)\nAttributable to:\nEquity holders of Barrick Gold Corporation\nNon-controlling interests (note 21)\n'>
<

## tabula-java

```java
// requires latest version of tabula-java built from source with PDFBox 3.x
// and saved as tabula.jar
// javac -cp "./tabula.jar" PDFGetter.java
// java -cp "./tabula.jar" PDFGetter.java

import technology.tabula.*;
import technology.tabula.extractors.SpreadsheetExtractionAlgorithm;
import java.io.InputStream;
import java.io.File;
import java.io.IOException;
import java.util.List;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.Loader;
public class PDFGetter {


    public void get() {
        try {
            PDDocument document = Loader.loadPDF(new File("~/tabula-java/src/test/resources/technology/tabula/table_report.pdf"));
            SpreadsheetExtractionAlgorithm sea = new SpreadsheetExtractionAlgorithm();
            PageIterator pi = new ObjectExtractor(document).extract();
            int pageNo = 0;
            while (pi.hasNext()) {
                // iterate over the pages of the document
                Page page = pi.next();
                if (pageNo < 3) {  // Only do the first page for now
                    List<Table> table = sea.extract(page);
                    // iterate over the tables of the page
                    for(Table tables: table) {
                        List<List<RectangularTextContainer>> rows = tables.getRows();
                        // iterate over the rows of the table
                        for (List<RectangularTextContainer> cells : rows) {
                            // print all column-cells of the row plus linefeed
                            for (RectangularTextContainer content : cells) {
                                // Note: Cell.getText() uses \r to concat text chunks
                                String text = content.getText().replace("\r", " ");
                                System.out.print(text + "|");
                            }
                            System.out.println();
                        }
                    }
                }
                pageNo = pageNo + 1;
            }
            System.out.println("The PageNo: " + pageNo);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void main(String []args)
    {
        PDFGetter pdf = new PDFGetter();
        pdf.get();
    }
}
```

## Marker
`marker_single /path/to/file.pdf /path/to/output/folder --batch_multiplier 2 --max_pages 1` 

### Consolidated Statements Of Income

| Barrick Gold Corporation                                                                                                                 | Three months ended   | Six months ended   |      |         |         |        |       |        |
|------------------------------------------------------------------------------------------------------------------------------------------|----------------------|--------------------|------|---------|---------|--------|-------|--------|
| (in millions of United States dollars, except per share data) (Unaudited)                                                                | June 30,             | June 30,           |      |         |         |        |       |        |
| 2015                                                                                                                                     |                      | 2014               | 2015 | 2014    |         |        |       |        |
| Revenue (notes 5 and 6)                                                                                                                  | $                    | 2,231              | $    | 2,458   | $ 4,476 | $      | 5,105 |        |
| Costs and expenses (income) Cost of sales (notes 5 and 7)                                                                                |                      | 1,689              |      | 1,631   |         | 3,397  |       | 3,350  |
| General and administrative expenses                                                                                                      | 70                   |                    | 82   |         | 137     |        | 185   |        |
| Exploration, evaluation and project expenses (note 8)                                                                                    |                      | 97                 |      | 105     |         | 183    |       | 205    |
| Impairment charges (note 10B)                                                                                                            |                      | 35                 |      | 512     |         | 40     |       | 524    |
| Loss on currency translation                                                                                                             |                      | 33                 |      | 31      |         | 31     |       | 110    |
| Closed mine rehabilitation                                                                                                               |                      | (19)               |      | 27      |         | (11)   |       | 49     |
| Loss (gain) on non-hedge derivatives (note 18D)                                                                                          |                      | 8                  |      | (44)    | 11      |        | (65)  |        |
| Other expense (note 10A)                                                                                                                 |                      | 32                 |      | 17      |         | 14     |       | 36     |
| Income before finance items and income taxes                                                                                             | 286                  |                    |      |         |         |        |       |        |
| $                                                                                                                                        | $                    | 97                 | $    | 674     | $       | 711    |       |        |
| Finance items Finance income                                                                                                             |                      | 2                  |      | 3       |         | 4      |       | 6      |
| Finance costs (note 11)                                                                                                                  |                      | (194)              |      | (200)   | (390)   |        | (401) |        |
| Income (loss) before income taxes                                                                                                        | 94                   |                    |      |         |         |        |       |        |
| $                                                                                                                                        | $                    | (100) $            | 288  | $       | 316     |        |       |        |
| Income tax expense (note 12)                                                                                                             |                      | (103)              |      | (123)   | (208)   |        | (412) |        |
| Net income (loss)                                                                                                                        | (9)                  |                    |      |         |         |        |       |        |
| $                                                                                                                                        | $                    | (223) $            | 80   | $       | (96)    |        |       |        |
| Attributable to: Equity holders of Barrick Gold Corporation                                                                              | $                    | (9)                | $    | (269) $ | 48      | $      | (181) |        |
| Non-controlling interests (note 21)                                                                                                      | $                    | -                  | $    | 46      | $       | 32     | $     | 85     |
| Earnings (loss) per share data attributable to the equity holders of Barrick Gold Corporation (note 9) Net income (loss)  Basic $ (0.01) | $                    | (0.23)             | $    | 0.04    | $       | (0.16) |       |        |
|                                                                                                                                          | Diluted              | $ (0.01)           | $    | (0.23)  | $       | 0.04   | $     | (0.16) |


## Unstructured

In [3]:
from unstructured.partition.auto import partition
elements = partition(pdf_file)

ModuleNotFoundError: No module named 'unstructured'

In [3]:
from collections import Counter

display(Counter(type(element) for element in elements))
print("")

Counter({unstructured.documents.elements.Text: 45,
         unstructured.documents.elements.NarrativeText: 6,
         unstructured.documents.elements.Footer: 3,
         unstructured.documents.elements.Title: 2})

In [4]:
display(*[(type(element), element.text) for element in elements[10:13]])

(unstructured.documents.elements.NarrativeText,
 'Six months ended June 30, 2014')

(unstructured.documents.elements.Text, '2015')

(unstructured.documents.elements.Text, '$')

In [5]:
print("\n\n".join([str(el) for el in elements]))

Consolidated Statements of Income

Barrick Gold Corporation

(in millions of United States dollars, except per share data) (Unaudited)

Revenue (notes 5 and 6) Costs and expenses (income) Cost of sales (notes 5 and 7) General and administrative expenses Exploration, evaluation and project expenses (note 8) Impairment charges (note 10B) Loss on currency translation Closed mine rehabilitation Loss (gain) on non-hedge derivatives (note 18D) Other expense (note 10A) Income before finance items and income taxes Finance items Finance income Finance costs (note 11) Income (loss) before income taxes Income tax expense (note 12) Net income (loss) Attributable to: Equity holders of Barrick Gold Corporation Non-controlling interests (note 21)

Earnings (loss) per share data attributable to the equity holders of Barrick Gold Corporation (note 9) Net income (loss) Basic Diluted

The accompanying notes are an integral part of these consolidated financial statements.

BARRICK SECOND QUARTER 2015

57


## Camelot

In [5]:
import camelot

tables = camelot.read_pdf(pdf_file, pages='1-end', flavor='lattice')
tables

<TableList n=0>

In [8]:
tables = camelot.read_pdf(pdf_file, flavor='stream')
print(tables[0])
print(tables[0].parsing_report)

<Table shape=(15, 6)>
{'accuracy': 95.02, 'whitespace': 28.89, 'order': 1, 'page': 1}


In [9]:
tables[0].df


,0,1,2,3,4,5
0,Barrick Gold Corporation,,Three months ended,,,Six months ended
1,"(in millions of United States dollars, except ...",,,"June 30,",,"June 30,"
2,,2015,,2014,2015,2014
3,Revenue (notes 5 and 6),"$ \n2,231",$,"2,458","$ \n4,476","$ \n5,105"
4,Costs and expenses (income),,,,,
5,Cost of sales (notes 5 and 7),"1,689",,"1,631","3,397","3,350"
6,General and administrative expenses,70,,82,137,185
7,"Exploration, evaluation and project expenses (...",97,,105,183,205
8,Impairment charges (note 10B),35,,512,40,524
9,Loss on currency translation,33,,31,31,110


## LlamaParse

### Consolidated Statements of Income

### Barrick Gold Corporation

(in millions of United States dollars, except per share data) (Unaudited)

|Description|Three months ended June 30, 2015|Three months ended June 30, 2014|Six months ended June 30, 2015|Six months ended June 30, 2014|
|---|---|---|---|---|
|Revenue (notes 5 and 6)|$ 2,231|$ 2,458|$ 4,476|$ 5,105|
|Costs and expenses (income)| | | | |
|Cost of sales (notes 5 and 7)|$ 1,689|$ 1,631|$ 3,397|$ 3,350|
|General and administrative expenses|$ 70|$ 82|$ 137|$ 185|
|Exploration, evaluation and project expenses (note 8)|$ 97|$ 105|$ 183|$ 205|
|Impairment charges (note 10B)|$ 35|$ 512|$ 40|$ 524|
|Loss on currency translation|$ 33|$ 31|$ 31|$ 110|
|Closed mine rehabilitation|($ 19)|$ 27|($ 11)|$ 49|
|Loss (gain) on non-hedge derivatives (note 18D)|$ 8|($ 44)|$ 11|($ 65)|
|Other expense (note 10A)|$ 32|$ 17|$ 14|$ 36|
|Income before finance items and income taxes|$ 286|$ 97|$ 674|$ 711|
|Finance items| | | | |
|Finance income|$ 2|$ 3|$ 4|$ 6|
|Finance costs (note 11)|($ 194)|($ 200)|($ 390)|($ 401)|
|Income (loss) before income taxes|$ 94|($ 100)|$ 288|$ 316|
|Income tax expense (note 12)|($ 103)|($ 123)|($ 208)|($ 412)|
|Net income (loss)|($ 9)|($ 223)|$ 80|($ 96)|
|Attributable to:| | | | |
|Equity holders of Barrick Gold Corporation|($ 9)|($ 269)|$ 48|($ 181)|
|Non-controlling interests (note 21)|$ 0|$ 46|$ 32|$ 85|
|Earnings (loss) per share data attributable to the equity holders of Barrick Gold Corporation (note 9)| | | | |
|Net income (loss)| | | | |
|Basic|($ 0.01)|($ 0.23)|$ 0.04|($ 0.16)|
|Diluted|($ 0.01)|($ 0.23)|$ 0.04|($ 0.16)|

The accompanying notes are an integral part of these consolidated financial statements.